protocols/autonat/src/lib.rs

```diff
@@ -0,0 +1,30 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the AutoNAT protocol.
+
+pub use self::behaviour::{Behaviour, Config};
+
+mod behaviour;
```

Comment:
```suggestion
pub mod behaviour;
```

Instead of exposing each type, I would suggest exposing the crate internal namespace structure, defining a types visibility at its definition. What do you think?

Also, in case I am not missing something, types like `AutoRetry` are unreachable as an external user.

Reply:
> Instead of exposing each type, I would suggest exposing the crate internal namespace structure, defining a types visibility at its definition. What do you think?

I guess its more a thing of personal preference. But since there are only few types, I prefer to have them on one layer instead of exposing the mod hierarchy inside of autonat; apart from that almost all types are in `behaviour` anyway. And other protocols mostly also directly expose the types, so it's more consistent with those as well.

Reply:
I am fine either way. The only thing I would like to prevent (which is already addressed in your code) is name repetition.

See https://github.com/libp2p/rust-libp2p/discussions/2174 and https://rust-lang.github.io/rust-clippy/master/index.html#module_name_repetitions

Also //CC @thomaseizinger as you were leading the discussion on naming conventions within rust-libp2p.

Reply:
My personal preference also goes for private modules and explicit re-exporting of types.

Public modules far too easily suffer from accidential API breaking changes, plus anything that is reachable from `lib.rs` is automatically excluded from rustc's dead-code analysis.

Naming looks good here.

Reply:
> plus anything that is reachable from lib.rs is automatically excluded from rustc's dead-code analysis.

Oh, good point. Thanks!